include/types.h

```diff
@@ -94,16 +147,57 @@ struct transfer_id {
 	friend bool operator!=(const transfer_id& lhs, const transfer_id& rhs) { return !(lhs == rhs); }
 };
 
-enum class error_policy {
-	ignore,
-	log_warning,
-	log_error,
-	panic,
+/// Instruction-graph equivalent of a USM pointer that permits pointer arithmetic.
+/// The offset will be applied by the executor once the allocation pointer is known.
+struct allocation_with_offset {
+	allocation_id id = null_allocation_id;
+	size_t offset_bytes = 0;
+
+	/// Constructs the equivalent of a null pointer.
+	allocation_with_offset() = default;
+
+	allocation_with_offset(const detail::allocation_id aid, const size_t offset_bytes = 0) : id(aid), offset_bytes(offset_bytes) {}
+
+	friend bool operator==(const allocation_with_offset& lhs, const allocation_with_offset& rhs) {
+		return lhs.id == rhs.id && lhs.offset_bytes == rhs.offset_bytes;
+	}
+	friend bool operator!=(const allocation_with_offset& lhs, const allocation_with_offset& rhs) {
+		return lhs.id == rhs.id && lhs.offset_bytes == rhs.offset_bytes;
+	}
 };
 
+inline allocation_with_offset operator+(const allocation_with_offset& ptr, const size_t offset_bytes) { return {ptr.id, ptr.offset_bytes + offset_bytes}; }
+
+inline allocation_with_offset& operator+=(allocation_with_offset& ptr, const size_t offset_bytes) {
+	ptr.offset_bytes += offset_bytes;
+	return ptr;
+}
+
```

Comment:
I thought of the `+` operator more like a fancy constructor for an `allocation_with_offset`. I could make it accept a `ptrdiff_t`, but then I would have to `static_cast` that at every point it's used, and there is currently no practical use for negative offsets in the code.

Reply:
I realize that absolute negative offsets make no sense, but what if someone e.g. builds an offset that is off by +10 and then adds -5 to that? that would be a valid offset, right?

In general, if the offset is stored as a `ptrdiff_t` you also have an opportunity to assert that it's not negative.

Reply:
I agree on principle that it makes sense to have signed sizes and offsets everywhere. But we faced the same problem in #201, where using signed types in box and region types would mean that we have to cast between signed and unsigned types everywhere, because SYCL (and to some extent the standard library) force us to be unsigned by default.

Reply:
I remember, but I don't think this applies in this case?

Reply:
I've removed the `operator+` "constructor" instead (yesterday's live discussion).